papis init: First-time usage is fragile and confusing

* papis version (`$ papis --version` or commit number): 0.13

# how to reproduce the issue
Command and output of `papis --log INFO <command>`:
If `~/Documents/papers` and `$XDG_CONFIG_HOME/papis/` don't exist:
(context for below: on my system, `$XDG_CONFIG_HOME=~/.local/etc/`)
```
No sections were found in the configuration file. Adding default ones (with a default library named 'papers')!
Traceback (most recent call last):
  File "/usr/bin/papis", line 33, in <module>
    sys.exit(load_entry_point('papis==0.13', 'console_scripts', 'papis')())
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/click/core.py", line 1685, in invoke
    super().invoke(ctx)
  File "/usr/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/site-packages/papis/commands/default.py", line 224, in run
    raise RuntimeError(
RuntimeError: Library 'papers' does not have any existing folders attached to it. Please define and create the paths in the configuration file
```

If `~/Documents/papers/` exists:
```
Creating configuration folder in '/home/gesh/.local/etc/papis'.
Creating scripts folder in '/home/gesh/.local/etc/papis/scripts'
No sections were found in the configuration file. Adding default ones (with a default library named 'papers')!
[WARNING] commands.init: Config file '/home/gesh/.local/etc/papis/config' already exists!
[WARNING] commands.init: This command may change some of its contents, e.g. whitespace and comments are not preserved.
```

In addition, trying to tab-complete to `papis init` has papis checking for a config file, failing, and emitting
```
Creating configuration folder in '/home/gesh/.local/etc/papis'.
Creating scripts folder in '/home/gesh/.local/etc/papis/scripts'
No sections were found in the configuration file. Adding default ones (with a default library named 'papers')!
```

This is unnecessarily brittle. A better flow could be:
- Tab-completion without a valid config file completes statically-knowable commands only (eg command-line arguments, builtin subcommands, etc)
- `papis init`, if it notices no libraries have been configured so far, prompts for a path to a first library, and offers to set it at default
- Under no circumstances should the flow of `papis init` create the config file before its check for the existence of a config file -- that's confusing and defeats the purpose of such a check.

OK, so testing:
> $ papis <TAB> Error while loading entrypoint 'EntryPoint(name='isbn',
> value='papis.isbn:Importer', group='papis.importer')': The 'arxiv>=1.0.0'
> distribution was not found and is required by papis.
> Error while loading entrypoint 'EntryPoint(name='isbn',
value='papis.isbn:explorer', group='papis.explorer')': The 'arxiv>=1.0.0'
distribution was not found and is required by papis.
> add        browse     config     exec       git        merge      rename      serve
> addto      cache      doctor     explore    init       mv         rm          update
> bibtex     citations  edit       export     list       open       run

The problem of `papis` creating the config directory, then warning that it might
overwrite the config it's already created still persists:
> $ echo "$XDG_CONFIG_HOME"
> /home/gesh/.local/etc
> $ rm -r ~/.local/etc/papis/
> $ papis init 
> Creating configuration folder in '/home/gesh/.local/etc/papis'.
> Creating scripts folder in '/home/gesh/.local/etc/papis/scripts'
> No sections were found in the configuration file. Adding default ones (with a default library named 'papers')!
> Error while loading entrypoint 'EntryPoint(name='isbn', value='papis.isbn:Importer', group='papis.importer')': The 'arxiv>=1.0.0' distribution was not found and is required by papis.
> Error while loading entrypoint 'EntryPoint(name='isbn', value='papis.isbn:explorer', group='papis.explorer')': The 'arxiv>=1.0.0' distribution was not found and is required by papis.
> [ERROR] commands.default: Library 'papers' does not have any existing folders attached to it. Please define and create the paths in the configuration file.
> [WARNING] commands.init: Config file '/home/gesh/.local/etc/papis/config' already exists!
> [WARNING] commands.init: This command may change some of its contents, e.g. whitespace and comments are not preserved.
> Do you want to continue? (Y/n): n

The error message could be improved, perhaps to
> Library '{lib}' has the following nonexistent folders attached to it.
> Please create them, or reconfigure '{lib}':
> {dir1}
> ...

---

In re your remarks on the design, perhaps have the documentation suggest
> papis config --edit

as an initial setup command, a la git?

In re `click` needing a valid `papis` configuration to be able to do anything,
perhaps the above suggestion could be a special case?
That is, at the top of the completion code, test if a valid configuration
exists. If so, pass control of completion to `click`.
Otherwise, complete to the above `papis config --edit`, perhaps even to
> papis config --edit # NOTE: Full completion needs a valid configuration

--- a/papis/commands/default.py
+++ b/papis/commands/default.py
@@ -175,7 +175,9 @@ def generate_profile_writing_function(profiler: "cProfile.Profile",
     help="Use number of processors for multicore functionalities in papis",
     type=str,
     default=None)
-def run(verbose: bool,
+@click.pass_context
+def run(ctx: click.Context,
+        verbose: bool,
         profile: str,
         config: str,
         lib: str,
@@ -232,10 +234,18 @@ def run(verbose: bool,
                 local_config_path,
                 papis.config.get_configuration())
     else:
-        logger.error(
-            "Library '%s' does not have any folders attached to it. Please "
-            "create and add the required paths to the configuration file.",
-            library)
+        config_file = papis.config.get_config_file()
+        if os.path.exists(config_file):
+            logger.error(
+                "Library '%s' does not have any folders attached to it. Please "
+                "create and add the required paths to the configuration file.",
+                library)
+        elif ctx.invoked_subcommand != "init":
+            logger.warning("No configuration file exists at '%s'.", config_file)
+            logger.warning("Create a configuration file and define your "
+                           "libraries before using papis. You can use "
+                           "'papis init /path/to/my/library' for a quick "
+                           "interactive setup.")
 
     # read in configuration from command-line
     sections = papis.config.get_configuration().sections()

--- a/papis/commands/default.py
+++ b/papis/commands/default.py
@@ -220,21 +220,22 @@ def run(verbose: bool,
     papis.config.set_lib_from_name(lib)
     library = papis.config.get_lib()
 
-    if not library.paths:
-        raise RuntimeError(
-            f"Library '{lib}' does not have any existing folders attached to it. "
-            "Please define and create the paths in the configuration file"
-            )
-
-    # Now the library should be set, let us check if there is a
-    # local configuration file there, and if there is one, then
-    # merge its contents
-    local_config_file = papis.config.getstring("local-config-file")
-    for path in library.paths:
-        local_config_path = os.path.expanduser(os.path.join(path, local_config_file))
-        papis.config.merge_configuration_from_path(
-            local_config_path,
-            papis.config.get_configuration())
+    if library.paths:
+        # Now the library should be set, let us check if there is a
+        # local configuration file there, and if there is one, then
+        # merge its contents
+        local_config_file = papis.config.getstring("local-config-file")
+        for path in library.paths:
+            local_config_path = os.path.expanduser(
+                os.path.join(path, local_config_file))
+            papis.config.merge_configuration_from_path(
+                local_config_path,
+                papis.config.get_configuration())
+    else:
+        logger.error(
+            "Library '%s' does not have any folders attached to it. Please "
+            "create and add the required paths to the configuration file.",
+            library)
 
     # read in configuration from command-line
     sections = papis.config.get_configuration().sections()

--- a/papis/commands/init.py
+++ b/papis/commands/init.py
@@ -100,9 +100,9 @@ def cli(dir_path: Optional[str]) -> None:
         if not papis.tui.utils.confirm("Do you want to continue?"):
             return
     else:
-        logger.info("Using config file: '%s'.", config_file)
+        logger.info("Initializing a new config file: '%s'.", config_file)
 
-    logger.info("Setting library location.")
+    logger.info("Setting library location:")
     dir_path = os.getcwd() if dir_path is None else dir_path
     library_name = papis.tui.utils.prompt(
         "Name of the library",

--- a/papis/commands/init.py
+++ b/papis/commands/init.py
@@ -104,11 +104,23 @@ def cli(dir_path: Optional[str]) -> None:
 
     logger.info("Setting library location:")
     dir_path = os.getcwd() if dir_path is None else dir_path
+
+    # FIXME: this is necessary because incorrectly configured libraries still
+    # show up in `papis.config.get_libs()`. We need better handling of missing
+    # libraries or ill-configured libraries.
+    known_libraries = []
+    for libname in papis.config.get_libs():
+        lib = papis.config.get_lib_from_name(libname)
+        if lib.paths and all(os.path.exists(path) for path in lib.paths):
+            known_libraries.append(libname)
+
     library_name = papis.tui.utils.prompt(
         "Name of the library",
         default=os.path.basename(dir_path),
-        bottom_toolbar="Known libraries: '{}'".format(
-            "', '".join(papis.config.get_libs())))
+        bottom_toolbar=(
+            "Known libraries: '{}'".format("', '".join(known_libraries))
+            if known_libraries else "No currently configured libraries")
+        )
 
     if library_name not in config:
         config[library_name] = {}

--- a/papis/commands/init.py
+++ b/papis/commands/init.py
@@ -127,13 +127,18 @@ def cli(dir_path: Optional[str]) -> None:
     local["dir"] = library_path
 
     logger.info("Setting library custom options.")
-    for setting, help_string in INIT_PROMPTS:
-        local[setting] = papis.tui.utils.prompt(
-            setting,
-            default=str(local.get(setting, defaults.get(setting))),
-            bottom_toolbar=help_string)
+    if papis.tui.utils.confirm("Want to add some standard settings?", yes=False):
+        for setting, help_string in INIT_PROMPTS:
+            local[setting] = papis.tui.utils.prompt(
+                setting,
+                default=str(local.get(setting, defaults.get(setting))),
+                bottom_toolbar=help_string)
 
     if papis.tui.utils.confirm("Do you want to save?"):
+        config_folder = papis.config.get_config_folder()
+        if not os.path.exists(config_folder):
+            os.makedirs(config_folder)
+
         with open(config_file, "w") as configfile:
             config.write(configfile)
 

--- a/papis/commands/init.py
+++ b/papis/commands/init.py
@@ -78,7 +78,7 @@ INIT_PROMPTS = [
 @click.argument(
     "dir_path",
     required=False,
-    type=click.Path(),
+    type=click.Path(file_okay=False),
     default=None,
     metavar="<LIBRARY DIRECTORY>",
     nargs=1,

--- a/papis/config.py
+++ b/papis/config.py
@@ -2,6 +2,8 @@ import os
 import configparser
 from typing import Dict, Any, List, Optional, Callable
 
+import click
+
 import papis.exceptions
 import papis.library
 import papis.logging
@@ -66,46 +68,36 @@ class Configuration(configparser.ConfigParser):
         for name in self["include"]:
             fullpath = os.path.expanduser(self.get("include", name))
             if os.path.exists(fullpath):
-                logger.debug("Including file '%s'.", name)
                 self.read(fullpath)
-            else:
-                logger.warning("'%s' not included because it does not exist.",
-                               fullpath)
+
+    def include_defaults(self) -> None:
+        for section in self.default_info:
+            self[section] = {}
+            for field in self.default_info[section]:
+                self[section][field] = self.default_info[section][field]
 
     def initialize(self) -> None:
         # ensure all configuration directories exist
         if not os.path.exists(self.dir_location):
-            logger.warning("Creating configuration folder in '%s'.", self.dir_location)
-            os.makedirs(self.dir_location)
-
-        if not os.path.exists(self.scripts_location):
-            logger.warning(
-                "Creating scripts folder in '%s'", self.scripts_location)
-            os.makedirs(self.scripts_location)
+            self.include_defaults()
+            return
 
         # load settings
         if os.path.exists(self.file_location):
-            logger.debug("Reading configuration from '%s'.", self.file_location)
             try:
                 self.read(self.file_location)
                 self.handle_includes()
             except configparser.DuplicateOptionError as exc:
-                logger.error("%s: %s", type(exc).__name__, exc, exc_info=exc)
+                click.echo("Failed to read configuration file "
+                           f"'{self.file_location}'.")
+                click.echo(f"Error: Duplicate option '{exc.option}' "
+                           f"in section {exc.section}")
                 raise SystemExit(1)
 
         # if no sections were actually read, add default ones
         if not self.sections():
-            logger.warning(
-                "No sections were found in the configuration file. "
-                "Adding default ones (with a default library named 'papers')!")
-
-            for section in self.default_info:
-                self[section] = {}
-                for field in self.default_info[section]:
-                    self[section][field] = self.default_info[section][field]
-
+            self.include_defaults()
             with open(self.file_location, "w") as configfile:
-                logger.info("Creating configuration file at '%s'.", self.file_location)
                 self.write(configfile)
 
         # ensure the general section and default-library exist in the config
@@ -115,17 +107,11 @@ class Configuration(configparser.ConfigParser):
                 libs[0] if libs else
                 self.default_info[GENERAL_SETTINGS_NAME]["default-library"])
 
-            logger.warning(
-                "No main '%s' section found in the configuration file. "
-                "Setting '%s' as the default library!",
-                GENERAL_SETTINGS_NAME, default_library)
-
             self[GENERAL_SETTINGS_NAME] = {"default-library": default_library}
 
         # evaluate the python config
         configpy = get_configpy_file()
         if os.path.exists(configpy):
-            logger.debug("Executing configuration script '%s'.", configpy)
             with open(configpy) as fd:
                 exec(fd.read())
 

--- a/papis/config.py
+++ b/papis/config.py
@@ -97,8 +97,6 @@ class Configuration(configparser.ConfigParser):
         # if no sections were actually read, add default ones
         if not self.sections():
             self.include_defaults()
-            with open(self.file_location, "w") as configfile:
-                self.write(configfile)
 
         # ensure the general section and default-library exist in the config
         if GENERAL_SETTINGS_NAME not in self:

--- a/papis/database/cache.py
+++ b/papis/database/cache.py
@@ -144,7 +144,7 @@ class Database(papis.database.base.Database):
             import pickle
             with open(cache_path, "rb") as fd:
                 self.documents = pickle.load(fd)
-        else:
+        elif self.get_dirs():
             logger.info("Indexing library. This might take a while...")
             folders: List[str] = sum(
                 [papis.utils.get_folders(d) for d in self.get_dirs()],
@@ -155,6 +155,9 @@ class Database(papis.database.base.Database):
                 self.maybe_compute_id(doc)
             if use_cache:
                 self.save()
+        else:
+            self.documents = []
+
         logger.debug("Loaded %d documents.", len(self.documents))
         return self.documents
 

